sequencing-server/src/utils/hasura.ts

```diff
@@ -0,0 +1,230 @@
+  }
+
+  throw new Error('Could not determine the user sending the request');
+}
+
+/**
+ * This function checks to see if request coming from the Hasura action is being executed
+ * by a user with enough permissions. To check this we need to call a GQL query with the
+ * action key and the user session. If we ever merge the databases this should be changed
+ * to call the db function that does the permission checking directly.
+ *
+ * Some endpoints don't need to be checked which is why we have the mappings of endpoints
+ * to actions to check.
+ *
+ * @param url The endpoint that we're checking the permissions for.
+ * @param graphqlClient The GQL Client we're using to make the permission check.
+ * @returns True if the user is able to call the given action, false otherwise.
+ */
+export async function canUserPerformAction(
+  url: string,
+  graphqlClient: GraphQLClient,
+  hasuraSession: AuthSessionVariables,
+  body: any,
+): Promise<boolean> {
+  const role = hasuraSession['x-hasura-role'];
+  const username = hasuraSession['x-hasura-user-id'];
+
+  // The aerie_admin role always has NO_CHECK permissions on all actions.
+  if (role === 'aerie_admin') {
+    return true;
+  }
+
+  const permissionCheckQuery = await graphqlClient.request(
+    gql`
+      query checkUserPermissions($role: user_roles_enum!, $key: String!) {
+        user_role_permission_by_pk(role: $role) {
+          action_permissions(path: $key)
+        }
+      }
+    `,
+    {
+      role,
+      key: ENDPOINTS_TO_ACTION_KEY[url],
+    },
+  );
+
+  const permission = permissionCheckQuery.user_role_permission_by_pk.action_permissions as HasuraPermissions | null;
+
+  if (permission === null) {
+    return false;
+  }
+
+  const missionModelId = body.input.missionModelId as number;
+  const simulationDatasetId = body.input.simulationDatasetId as number;
+  let planId = body.input.planId as number;
+
+  // If we have a simulationDatasetId we need to get the planId to do permission checking.
+  if (simulationDatasetId !== null) {
+    planId = await getPlanId(graphqlClient, simulationDatasetId);
+  }
+
+  switch (permission) {
+    case HasuraPermissions.NO_CHECK:
+      return true;
+    case HasuraPermissions.MISSION_MODEL_OWNER:
+      return missionModelId !== null && (await isMissionModelOwner(graphqlClient, username, missionModelId));
+    case HasuraPermissions.OWNER:
+      return planId !== null && (await isPlanOwner(graphqlClient, username, planId));
+    case HasuraPermissions.PLAN_OWNER:
+      return planId !== null && (await isPlanOwner(graphqlClient, username, planId));
+    case HasuraPermissions.PLAN_COLLABORATOR:
+      return planId !== null && (await isPlanCollaborator(graphqlClient, username, planId));
+    case HasuraPermissions.PLAN_OWNER_COLLABORATOR:
+      return (
+        (planId !== null && (await isPlanOwner(graphqlClient, username, planId))) ||
+        (await isPlanCollaborator(graphqlClient, username, planId))
+      );
+  }
```

Comment:
The current switch statement is insufficient. Something roughly like this will satisfy the valid keys for all endpoints.
```ts
  switch (permission) {
    case HasuraPermissions.NO_CHECK:
      return true;
    case HasuraPermissions.MISSION_MODEL_OWNER:
      if (missionModelId !== null) {
         return await isMissionModelOwner(graphqlClient, username, missionModelId);
      }
      return await isMissionModelOwnerFromPlanId(graphqlClient, username, planId);
    case HasuraPermissions.OWNER:
      if ( ENDPOINTS_TO_ACTION_KEY[url] == create_expansion_rule 
           || ENDPOINTS_TO_ACTION_KEY[url] == create_expansion_set ) {
            return await isMissionModelOwner(graphqlClient, username, missionModelId);
      }
      return planId !== null && (await isPlanOwner(graphqlClient, username, planId));
    case HasuraPermissions.PLAN_OWNER:
      if ( ENDPOINTS_TO_ACTION_KEY[url] == create_expansion_rule 
           || ENDPOINTS_TO_ACTION_KEY[url] == create_expansion_set ) {
            return await isAnyPlanOwner(graphqlClient, username, missionModelId);
      }
      return planId !== null && (await isPlanOwner(graphqlClient, username, planId));
    case HasuraPermissions.PLAN_COLLABORATOR:
      if ( ENDPOINTS_TO_ACTION_KEY[url] == create_expansion_rule 
           || ENDPOINTS_TO_ACTION_KEY[url] == create_expansion_set ) {
            return await isAnyPlanCollaborator(graphqlClient, username, missionModelId);
      }
      return planId !== null && (await isPlanCollaborator(graphqlClient, username, planId));
    case HasuraPermissions.PLAN_OWNER_COLLABORATOR:
       if ( ENDPOINTS_TO_ACTION_KEY[url] == create_expansion_rule 
           || ENDPOINTS_TO_ACTION_KEY[url] == create_expansion_set ) {
            return (await isAnyPlanOwner(graphqlClient, username, missionModelId)) 
                    ||  (await isAnyPlanCollaborator(graphqlClient, username, missionModelId));
      }
      return (
        (planId !== null && (await isPlanOwner(graphqlClient, username, planId))) ||
        (await isPlanCollaborator(graphqlClient, username, planId))
      );
  }
```

Reply:
I implemented this in a slightly different way which is hopefully more clear and still gets each case.